### PR TITLE
Implement method-level test dependencies

### PR DESCRIPTION
**Pull request by stevenaw**
Relates to https://github.com/nunit/nunit/pull/5365
Contributes to #51 

Builds on the previous PR to adds support for method->method dependencies. Some logic has been moved out of `WorkItemBuilder` into a dedicated `TestDependencyManager` to allow for fixture-specific and method-specific implementations of building the dependency tree. The rest of the dependency management should now be fixture-agnostic, including changes to `MarkInvalidDependenciesAsInvalid` now.

All fixture test scenarios have been duplicated as method-focused implementations. New tests were also added to attribute misuse (attempting method->fixture or fixture->method dependencies was previously identified as undesirable)

### CODE REVIEW

File: src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace NUnit.Framework
     /// <summary>
     /// Specifies that a test fixture must run after another fixture.
     /// </summary>
-    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs:12]: I think it might be cleaner to split this into two attributes:
`DependsOnFixture(Type)` and `DependsOnMethod(string)`
You then don't have to test if the Type constructor is used on a method and vice versa.

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.cs [hunk 2]
```diff
@@ -28,307 +18,10 @@ public void AttributeUsageIsClassLevelSingleUse()
 
             using (Assert.EnterMultipleScope())
             {
-                Assert.That(usage.ValidOn, Is.EqualTo(AttributeTargets.Class));
+                Assert.That(usage.ValidOn, Is.EqualTo(AttributeTargets.Class | AttributeTargets.Method));
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.cs:25]: You changed test which no longer matches the test name.
Irrelevant if you concur and split the attribute.

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                return null;
+
+            var dependencyGraph = new Dictionary<ITest, ITest>(parent.Tests.Count);
+
+            foreach (var child in parent.Tests)
+            {
+                if (child is not Test test || !dependencyByMethod.TryGetValue(test, out string? dependencyMethod))
+                    continue;
+
+                if (testsByName.TryGetValue(dependencyMethod, out Test? dependencyTest))
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs:156]: The `ITest.Name` might not match `methodName` if either the user sets the `Name` of the test manually or if the test has parameters.

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                {
+                    // Check for self-referential base dependency first
+                    if (test.Method?.MethodInfo.DeclaringType is Type testType &&
+                        dependencyTest.Method?.MethodInfo.DeclaringType is Type dependencyTestType &&
+                        testType != dependencyTestType &&
+                        testType.IsSubclassOf(dependencyTestType))
```
Comment (manfred-brands) [inline src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs]: If _testType_ is derived from _dependencyTestType_ we have a dependency on a base class test, that should not be an issue. A TestFixture flattens the hierarchy.

I think the test should have been:

```suggestion
                       !testType.IsSubclassOf(dependencyTestType))
```

This also means we need 2 more tests:
1. Where the dependency method is in a completely different class. This should fail.
2. Where the dependency method is in the base class. This should pass.

Maybe also add an `abstract` test in a base class with implementation in the derived class.
This is so the base class can do all tests once the derived class this some initial work (equivalent to SetUp but a test)

File: src/NUnitFramework/testdata/DependsOnAttributeFixture.Fixtures.cs
Comment (manfred-brands) [inline src/NUnitFramework/testdata/DependsOnAttributeFixture.Fixtures.cs:1]: Should this file be renamed to match the one on method and the naming of the Test file?

**Review by manfred-brands — CHANGES_REQUESTED** [review]
@stevenaw Nice, most changes are in the test. The sharing of the TestDependencyManager is great.

I think there is one bug, where the method dependency is in the base class. See separate comment.

Also see the remark on `Test.Name` vs `Test.Method.MethodInfo..Name`
You might want to add a parameterized test with `TestCase` and a dependency.
All Test created from these should have the dependency.

I don't think we should allow dependencies on a parameterized test.
If you agree we need code to detect this and tests.

File: src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs (hunk 1 above)
Comment (stevenaw) [reply src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs:12]: Interesting idea and I see where you're coming from with shifting this to a compile-time check. I'd like to keep this as a single attribute though, in order to ease adoption to the newer model. Previous discussion had focused on using a roslyn analyzer to help shift this to compile-time which would help even if not used by all. I've just been waiting to file the issue on the analyzer repo until after this PR landed

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.cs (hunk 2 above)
Comment (stevenaw) [reply src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.cs:25]: Thanks for catching this. I'll update it

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs (hunk 3 above)
Comment (stevenaw) [reply src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs:156]: > if either the user sets the Name of the test manually

Good point. I admit I had originally intended to keep this focused on method names for simplicity since there's a greater guarantee there will only ever be a single instance of the name within a suite. Customized test names _may_ appear several times. Though I see it could also lead to surprises if a method name-based setup is changed later to introduce a customized test name.

UPDATE: Actually, I see explicitly pursuing method-name-only can also cause some complications. I'll remove explicit references to method names, but _customized_ test names are really only supported for parameterized tests which I'd like to defer

> or if the test has parameters.

I'll have to confirm, but I think there's a lot of hidden complexity here as I think NUnit will treat a set of parameterized tests as a separate composite work item. I think it's also possible we also chose to remove the functionality you're describing for `Order` when we removed the code to apply the attribute's properties to child tests.

@manfred-brands @OsirisTerje 

One of the reasons previous attempts at this behaviour stalled out is that there's a number of different ways to approach these cases and the team had a hard time pinning down a minimum feature set. Both of the ideas described above can be implemented later as enhancements without changing the functionality we can choose to support today. Are we in agreement to make them dedicated follow-up issues? The parameterized method one in particular might warrant a dedicated PR to ensure we cover all the edge cases.

File: src/NUnitFramework/testdata/DependsOnAttributeFixture.Fixtures.cs
Comment (stevenaw) [reply src/NUnitFramework/testdata/DependsOnAttributeFixture.Fixtures.cs:1]: Oh, thanks for pointing that out. I thought I had done that

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs (hunk 4 above)
Comment (stevenaw) [reply src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs]: Thanks, good point. I only have a test for "SelfReferentialBaseClassMarksTestAsNotRunnable" but not for happy-path

Comment (stevenaw) [reply src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs]: Oh. Wait a minute. Yes, I think the original intend behind this test changed by accident. "SelfReferential" was supposed to be unhappy-path. I'll take a closer look these cases

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs (hunk 3 above)
Comment (stevenaw) [reply src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs:156]: > or if the test has parameters.

Oops. Sorry. I read this as if you were asking to _support_ parameterized tests but I see above that you've suggested otherwise

> I don't think we should allow dependencies on a parameterized test.
> If you agree we need code to detect this and tests.

I do agree 100%, at least as an initial version. Parameterized tests are at a different point in the hierarchy (different compositeworkitem) and it becomes much more complex to support. This might be an interesting follow-up enhancement for a later time as the solution to make it work also likely would allow _parallel_ start->finish dependencies to work.

**Comment by stevenaw** [comment]
Thanks for your review @manfred-brands 

I only have small windows of time right now, but I've pushed changes for all but two pieces of feedback. Of the remaining:

> Splitting this into two attributes

I'd prefer not to split this into separate attributes. I suspect it's simpler to reason about this as a single attribute and, while imperfect, we could still warn against misuse at compile time using the roslyn analyzer approach discussed previously. I suspect it might also feel more consistent with other attributes like `Parallelizable` and `Order` which use a single attribute for common functionality at fixture and method level.

> Explicitly not supporting parameterized tests

I misunderstood an earlier comment of yours, but I agree 100% with this. I have run out of time to add these tests right now but hope to very soon.

**Comment by stevenaw** [comment]
I was pleasantly surprised when I came back to this this evening that parameterized methods "just work". This works because of three key points:

1. We usually apply test properties to parameterized methods at the method grouping level unless exposed as properties on the attribute which sets up the parameterized suite itself. In this case this is a distinct attribute and so the whole parameterized method suite is marked as a dependent rather than the individual contained test cases.
2. This allows the suite of parameterized tests to start, finish, and record a result as a distinct unit
3. Our previous decision not to support parallelism means the parameterized suite must finish as a unit either before or after the the test on which it may depend

I also don't think it would complicate any potential future parallelism support to add this in since it's all still tracked as sibling nodes (parameterized method suite -> test method) rather than passing the dependency _itself_ down into the child tests of the parameterized method suite.

File: src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs (hunk 1 above)
Comment (manfred-brands) [reply src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs:12]: Having analyzers addressing issues in existing code is one thing, designing new code so that it needs an analyzer is another.

I much rather write an analyzer and code fix to replace `Order` with `DependsOnMethod` to help the adoption.

Here the two would have different functions, one for dependencies between fixtures, the other for dependency within a fixture.

The reason we have an analyzer checking types of _actual_ and _expected_ is because the API treated everything as object until execution time, allowing comparing apples with oranges.

Comment (manfred-brands) [reply src/NUnitFramework/framework/Attributes/DependsOnAttribute.cs:12]: In all your test code you treat them as completely separate cases.
There is no case which uses both, 2 fixtures, one depending on the other and having method dependencies internally.

File: src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    [DependsOn(nameof(FixtureMethod))]
+    public class MethodDependencyInvalidStringOnFixture
+    {
+        [Test]
+        public void FixtureMethod()
+        {
+            Assert.That(true, Is.True);
+        }
+    }
+
```
Comment (manfred-brands) [inline src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs]: This is tested in the `DependsOnAttributeTests.Fixture.cs` because it is an attribute on a fixture.
Should the code here then also be moved to `DependsOnAttributeFixture.Fixture.cs`?

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+                {
+                    Assert.That(dependencyTargetResult.ResultState.Status, Is.EqualTo(TestStatus.Passed));
+                    Assert.That(dependantMethodResult.ResultState.Status, Is.EqualTo(TestStatus.Failed));
+                    Assert.That(dependantMethodResult.ResultState.Label, Is.EqualTo("Invalid"));
+                    Assert.That(dependantMethodResult.Message, Does.Contain("DependsOnAttribute Type constructor may only be used on fixtures."));
+                }
+            }
+
+            [Test]
+            public void ParameterizedTestDependsOnRegularTestIsMarkedInvalid()
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs]: The problem with descriptive names indicating the outcome is that those names need to change if you change the outcome.

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+                var work = TestBuilder.CreateWorkItem(typeof(MethodCantDependOnParameterizedTest));
+                var result = TestBuilder.ExecuteWorkItem(work);
+
+                var beforeResult = result.Children.Single(x => x.Name == nameof(MethodCantDependOnParameterizedTest.B));
+                var afterResult = result.Children.Single(x => x.Name == nameof(MethodCantDependOnParameterizedTest.ParameterizedTestB));
+
+                using (Assert.EnterMultipleScope())
+                {
+                    Assert.That(beforeResult.ResultState.Status, Is.EqualTo(TestStatus.Passed));
+                    Assert.That(afterResult.ResultState.Status, Is.EqualTo(TestStatus.Passed));
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs:342]: This only asserts that both are run, not that they are run in the required order.
Like the name, it is a result of expecting failure, then order doesn't matter.
Now that the succeed the order does matter.

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+                var work = TestBuilder.CreateWorkItem(typeof(MethodCantDependOnParameterizedTest));
+                var result = TestBuilder.ExecuteWorkItem(work);
+
+                var beforeResult = result.Children.Single(x => x.Name == nameof(MethodCantDependOnParameterizedTest.ParameterizedTestA));
+                var afterResult = result.Children.Single(x => x.Name == nameof(MethodCantDependOnParameterizedTest.A));
+
+                using (Assert.EnterMultipleScope())
+                {
+                    Assert.That(beforeResult.ResultState.Status, Is.EqualTo(TestStatus.Passed));
+                    Assert.That(afterResult.ResultState.Status, Is.EqualTo(TestStatus.Passed));
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs:361]: We need a check that all parameterized subtest of `ParameterizedTestA` run before `A`

File: src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        [Test]
+        [DependsOn(nameof(BaseMethod))]
+        public void Self()
+        {
+            Assert.That(true, Is.True);
+        }
+    }
+
+    [TestFixture]
+    public class MethodCantDependOnParameterizedTest
```
Comment (manfred-brands) [inline src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs]: Name not matching asserted outcome:

```suggestion
    public class MethodDependOnParameterizedTest
```

**Review by manfred-brands — CHANGES_REQUESTED** [review]
It is great that parameterized test, because they are grouped, work out of the box!

As a result of this unexpected outcome, a few test classes and test have the wrong outcome in their name.

File: src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs (hunk 3 above)
Comment (OsirisTerje) [reply src/NUnitFramework/framework/Internal/Execution/TestDependencyManager.cs:156]: Imho, I am always in favor of moving forward with incomplete features.  This is a new featureset, and even less reason to worry.  If we get a series of bug reports and suggestions this way or another, it just means people are interested.  So, imho - run man, run!

**Comment by stevenaw** [comment]
THanks @manfred-brands . I think I've covered the requested scenarios. Let me know what you think

File: src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs [hunk 10]
```diff
@@ -292,35 +281,60 @@ public void Self()
...
+            FixtureDependencyEvents.Record();
             Assert.That(true, Is.True);
         }
     }
 
+    [TestFixture]
+    public class MethodDependsOnParameterizedTestWithFailure
+    {
+        [Test]
+        [DependsOn(nameof(A))]
```
Comment (manfred-brands) [inline src/NUnitFramework/testdata/DependsOnAttributeFixture.Methods.cs]: I presume this should have been `nameof(ParameterizedTestA)`

File: src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs [hunk 11]
```diff
@@ -325,34 +325,54 @@ public void TypeConstructorOnMethodMarksMethodInvalid()
...
+
+                var beforeResult = result.Children.Single(x => x.Name == nameof(MethodDependsOnParameterizedTestWithFailure.ParameterizedTestA));
+                var afterResult = result.Children.Single(x => x.Name == nameof(MethodDependsOnParameterizedTestWithFailure.A));
+
+                using (Assert.EnterMultipleScope())
+                {
+                    Assert.That(beforeResult.ResultState.Status, Is.EqualTo(TestStatus.Failed));
+                    Assert.That(beforeResult.ResultState.Site, Is.EqualTo(FailureSite.Child));
+                    Assert.That(afterResult.ResultState.Status, Is.EqualTo(TestStatus.Failed));
+                    Assert.That(afterResult.ResultState.Label, Is.EqualTo("Invalid"));
```
Comment (manfred-brands) [inline src/NUnitFramework/tests/Attributes/DependsOnAttributeTests.Methods.cs]: This is the wrong status, test `A` doesn't fail, but should have been `Skipped`.
Due to the wrong name in the `DependsOn` it fails now.

If I add a check on the reason:
```csharp
Assert.That(afterResult.Message, Does.Contain("Dependency on ParameterizedTestA did not pass"));
```

The code fails with:
```text
 Assert.That(afterResult.Message, Does.Contain("Dependency ParameterizedTestA did not pass"))
  Expected: String containing "Dependency on ParameterizedTestA did not pass"
  But was:  "Circular or self-referential test dependency detected."
```

**Review by manfred-brands — COMMENTED** [review]
@stevenaw One test was wrong, see below.

I also belated saw a typo in `FileCycleReason` instead of `FailCycleReason`

I have pushed two commits to address both, to shorten the cycle.

Looks good now.
Except for the disagreement between us about splitting `DependsOnAttribute`.
We need an arbiter @OsirisTerje, could you please look at the comments on this attribute?

**Comment by OsirisTerje** [comment]
> Except for the disagreement between us about splitting DependsOnAttribute.
We need an arbiter @OsirisTerje, could you please look at the comments on this attribute?

When I look at this there are two different "things" here, one is a dependency between fixtures, the other is a dependency between methods (in a fixture).   
Merging them into one attribute - it worries me because they are different, and a single attribute means compromises, and to me, it is more unclear and less typesafe. 

## Comparison

| |Common Attribute | Split attributes | 
|---|---|---|
|**Discoverability**| One name to type; IntelliSense shows one entry with two overloads that look identical until you check parameter types | Two names; typing `DependsOn` in IntelliSense surfaces both immediately, and the name itself tells you which one applies before you even read the signature | |
|**Typesafety -user make wrong choice** | `[DependsOn(typeof(X))]` moved from a fixture onto a method compiles fine, fails silently until test run | `[DependsOnFixture(typeof(X))]` on a method fails to compile — the mistake never ships |
| **Conceptual model** | "One thing: depends-on" — mirrors MbUnit's original single-attribute precedent from issue #51, but that is way back in time | Two distinct concepts made explicit: cross-fixture ordering vs. intra-fixture ordering, which *are* semantically different (different scopes, different failure messages, different edge cases ref. parameterized-tests ) |
| **Misuse caught by...** | Runtime (`MakeInvalid` at discovery), or later an analyzer rule we have to write | The C# compiler, via `AttributeUsage(Class)` / `AttributeUsage(Method)` —obvious to do that. |

## Splitting

We have two different scopes,  a fixture dependency points to other fixtures, a method dependency points to other methods in the same fixture,  cross-dependencies are not allowed.

So these two scopes are completely asymmetrical, which is another argument for splitting.  One attribute means the *reach* of the dependency flips depending on which constructoir you use.  Nothing about the attribute name signals "this one only sees methods in your own fixture, that one sees fixtures in the whole assembly.".  Two attributes with appropriate names makes this more obvious, and setting the AttributeUSage helps too.

## Costs

One attribute :  Have to write an analyzer rule later
Two attributes:  Slightly more surface area (two doc pages, possibly minor duplication like `AllowFailure`), but that's an authoring cost, not a user-safety cost

## Comments

>   >Splitting this into two attributes

>I'd prefer not to split this into separate attributes. I suspect it's simpler to reason about this as a single attribute and, while imperfect, we could still warn against misuse at compile time using the roslyn analyzer approach discussed previously. I suspect it might also feel more consistent with other attributes like Parallelizable and Order which use a single attribute for common functionality at fixture and method level.

I am not worried about having to match earlier attributes (Order/Parallelizable). This/these new attribute(s) are completely new, and I believe we should have them as typesafe as we can, and not introduce.

About *simpler to reason about*, I see the point here, but we could still describe them as belonging to a DependsOn group (in the documentation).  It might also be harder to explain that it has a dual nature , instead of just being explicit about that. So I actually find two attributes "simpler to reason about" 😄 .

Also, splitting the public attribute isn't a redesign, it's exposing a distinction that already exists internally.  They do use common code, but the attribute DependsOn looks like it have two different responsibilities, and it also expose nullable Type and string properties to handle this.  

Like @manfred-brands said, I do like the  common use of the TestDependencyManager, that is great. It also looks to me that there is a lot of more common code there. If that's true, it means splitting doesn't seem to affect more than the attribute itself, very little else need to change, and that change seems to me to tighten the code more.

**Comment by stevenaw** [comment]
Ok thanks @OsirisTerje     @manfred-brands 
Can we do that in a follow-up PR - still obviously within 5.0.

I've had a very hard time keeping up with mid-PR design changes in each of these dependency-related PRs and I would really prefer to tackle this one independently. I suspect most of the changes will effectively surface as a "rename" in the attribute targeted, which will likely also be easier to review if done independently.

EDIT:
Also, for the non-fixture dependency, do we prefer "DependsOnMethod" or "DependsOnTest" ? A method is a .NET construct while a fixture and a test are NUnit constructs. There is also a nice parity between attributes `TestFixture` -> `DependsOnFixture` and corresponding `Test` and `DependsOnTest`

**Comment by OsirisTerje** [comment]
Should we then just merge this, and pick up the strays in an upcoming PR ?

I think that might be easier.

**Comment by stevenaw** [comment]
@OsirisTerje I think that'd be my preference if it's alright with @manfred-brands

**Comment by manfred-brands** [comment]
> @OsirisTerje I think that'd be my preference if it's alright with @manfred-brands

As the DependsOnAttribute is already in master, I would say merge this PR now and create a new one splitting the attribute.

I like your `DependsOnFixture` and `DependsOnTest` naming and reasoning, so go for that.

**Review by manfred-brands — APPROVED** [review]
Approved for merge, with follow-up PR to split the attribute.
